File: backend/app/ChessLogic/MatchDatas/DataStores/MatchState.php
<?php

namespace ChessLogic\MatchDatas\DataStores;

use ChessLogic\ChessBoard\ChessBoardFactory;
use ChessLogic\ChessBoard\ChessPieces\Definitions\Side;
use ChessLogic\ChessBoard\ChessPieces\Pieces\ChessPiece;
use ChessLogic\Notations\PieceNotations\SANNotationRow;

use JsonSerializable;

class MatchState implements JsonSerializable
{
    public ?string $MatchDuration;

    /** @var ChessPiece[][] */
    public array $PieceMatrix;

    public Side $CurrentSide = Side::White;

    /** @var SANNotationRow[] */
    public array $Notations = [];

    public bool $HasWKMoved = false;
    public bool $HasWRAMoved = false;
    public bool $HasWRHMoved = false;
    public bool $HasBKMoved = false;
    public bool $HasBRAMoved = false;
    public bool $HasBRHMoved = false;

    /** @var array{0:int,1:int}|null */
    public ?array $EnPassantTarget = null;

    public function __construct()
    {
        $this->PieceMatrix = ChessBoardFactory::boardFiller();
    }
       
    public function jsonSerialize() : array
    {
        return [
            'matchDuration' => $this->MatchDuration,
            'currentSide' => $this->CurrentSide->value,
            
            'pieceMatrix' => array_map(fn($row) => array_map(
                    fn($piece) => $piece ? $this->pieceToString($piece) : null,
                    $row
                ),
                $this->PieceMatrix
            ),

            'notations' => array_map(fn($notation) => $this->notationsToArray($notation),
                $this->Notations
            ),

            'hasWKMoved' => $this->HasWKMoved,
            'hasWRAMoved' => $this->HasWRAMoved,
            'hasWRHMoved' => $this->HasWRHMoved,
            'hasBKMoved' => $this->HasBKMoved,
            'hasBRAMoved' => $this->HasBRAMoved,
            'hasBRHMoved' => $this->HasBRHMoved,

            'enPassantTarget' => $this->EnPassantTarget,
        ];
    }

    private function pieceToString(ChessPiece $piece) : string
    {
        return $piece->Name->value . $piece->Color->value;
    }

    private function notationsToArray(SANNotationRow $row) : array
    {
        return [
            'round' => $row->Round,
            'white' => $row->WhitesNotation,
            'black' => $row->BlacksNotation,
        ];
    }

    public static function fromArray(array $data) : MatchState
    {
        $state = new MatchState();

        $state->MatchDuration = $data['matchDuration'] ?? '3 | 2';
        $state->CurrentSide = Side::from($data['currentSide']);
        $state->HasWKMoved = $data['hasWKMoved'] ?? false;
        $state->HasWRAMoved = $data['hasWRAMoved'] ?? false;
        $state->HasWRHMoved = $data['hasWRHMoved'] ?? false;
        $state->HasBKMoved = $data['hasBKMoved'] ?? false;
        $state->HasBRAMoved = $data['hasBRAMoved'] ?? false;
        $state->HasBRHMoved = $data['hasBRHMoved'] ?? false;
        $state->EnPassantTarget = $data['enPassantTarget'] ?? null;

        $state->PieceMatrix = array_map(fn($row) => array_map(
                fn($piece) => $piece ? ChessPiece::fromString($piece) : null,
                $row
            ),
            $data['pieceMatrix']
        );

        $state->Notations = array_map(fn($row) => SANNotationRow::fromArray($row),
            $data['notations'] ?? []
        );

        return $state;
    }
}

File: NeuChessHu_app/NeuChessHu/ChessMechanics/MatchData/MatchDatas/Models/MatchState.cs
using ChessMechanics.ChessBoard.ChessPieces;
using ChessMechanics.ChessBoard.Definitions;
using ChessMechanics.Common;
using ChessMechanics.MatchData.MatchDatas.Models.DomainModels;
using System.Collections.ObjectModel;

namespace ChessMechanics.MatchData.MatchDatas.Models;

public class MatchState : ObservableBase
{
    ChessPiece[,]? pieceMatrix;
    Side currentSide;
    string? matchDuration;

    public string? MatchID { get; internal set; }

    public string MatchDuration
    {
        get => matchDuration!;
        set { matchDuration = value; RaisePropertyChanged(); }
    }

    public ChessPiece[,] PieceMatrix
    {
        get => pieceMatrix!;
        set { pieceMatrix = value; RaisePropertyChanged(); }
    }

    public Side CurrentSide
    {
        get => currentSide;
        internal set { currentSide = value; RaisePropertyChanged(); }
    }

    public ObservableCollection<SANNotationRow> Notations { get; internal set; } = [];

    public bool HasWKMoved { get; internal set; }
    public bool HasWRAMoved { get; internal set; }
    public bool HasWRHMoved { get; internal set; }
    public bool HasBKMoved { get; internal set; }
    public bool HasBRAMoved { get; internal set; }
    public bool HasBRHMoved { get; internal set; }

    public Tuple<int, int>? EnPassantTarget { get; internal set; }

    internal static MatchState CreateMatchState() => new();
}